cloud/openstack/os_nova_quotas.py

```diff
@@ -0,0 +1,119 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this software.  If not, see <http://www.gnu.org/licenses/>.
+
+try:
+    import shade
+    HAS_SHADE = True
+except ImportError:
+    HAS_SHADE = False
+
+
+DOCUMENTATION = '''
+---
+module: os_nova_quotas
+short_description: Update/Delete nova quotas in a OpenStack project.
+extends_documentation_fragment: openstack
+author: "Ghe Rivero (@GheRivero)"
+version_added: "2.1"
+description:
+   - Update or delete quotas in a OpenStack project.
+options:
+  state:
+    description:
+      - Should the resource be present or absent
+    choices: ['present', 'absent']
+    required: false
+    default: present
+  quota:
+    description:
+      - Dictionary containing the quotas and values you want to update
+    required: false
+  project:
+    description:
+      - The name or id of the project to which the quota should be assigned
+    required: true
+
+requirements: ["shade"]
+'''
+
+EXAMPLES = '''
+# Update the quota of the project 'mycloud' to 42 cores and 20 instances
+- os_nova_quotas:
+    project=mycloud
+    state=present
+    quota:
+      cores: 42
+      instances: 20
+
+# Delete quotas for the project 'mycloud' and back to defaults:
+- os_nova_quotas:
+    cloud=mycloud
+    state=absent
+    project: demo
+'''
+
+RETURN = '''
+quotas:
+    description: Dictionary listing all the compute quotas and their values for the project
+    returned: On success
+    type: dictionary
+
+'''
+
+EXCLUDE_KEYS = ["HUMAN_ID", "human_id", "NAME_ATTR", "request_ids", "x_openstack_request_ids" ]
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        project = dict(required=False),
+        quota = dict(default=None, type='dict'),
+        state = dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec, **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+
+    try:
+        cloud = shade.operator_cloud(**module.params)
+
+        old_quotas = cloud.get_compute_quotas(module.params['project'])
+
+        if module.params['state'] == 'present':
+            cloud.set_compute_quotas(module.params['project'],
+                                    **module.params['quota'])
+
+        elif module.params['state'] == 'absent':
+            cloud.delete_compute_quotas(module.params['project'])
+
+        new_quotas = cloud.get_compute_quotas(module.params['project'])
+
+        map(old_quotas.pop, EXCLUDE_KEYS)
+        map(new_quotas.pop, EXCLUDE_KEYS)
+
+        module.exit_json(changed=old_quotas != new_quotas, quotas=new_quotas)
+
+    except shade.OpenStackCloudException as e:
+        module.fail_json(msg=str(e), extra_data=e.extra_data)
```

Comment:
Do not include extra_data. That's a hold-over from old code. Everything should be present in the str(e) conversion.